VDA-SBT Updated for new VDA-Verification-Contract

ITStar10 opened the pull request:
Updated
	parameter type of claimSBT() function
	proof generation code updated in the contract & TS test code

ITStar10 committed:
VDA-SBT Updated for new VDA-Verification-Contract

Updated
	parameter type of claimSBT() function
	proof generation code updated in the contract & TS test code

tahpot commented on VDA-SBT/contracts/SoulboundNFT.sol:
```diff
@@ -190,57 +164,51 @@ contract SoulboundNFT is VDAVerificationContract,
-            verifyRequest(did, _trustedAddresses, params, signature, proof);
+                _nonce[did]);
+            // verifyRequest(did, _trustedAddresses, params, signature, proof);
+            verifyData(params, signature, proof);
```
This should be `verifyRequest()` as we need to prevent replay attacks.

tahpot reviewed (changes requested):
@ITStar10 This needs a lot of changes with the new VDAVerificationContract:

1. Test needs to generate signed requests and the smart contract needs to verify with `verifyRequest()`. This will require including the `proof` from the DID Document of the DID that is submitting the request
2. Test is correctly generating a proof for the SBT, however that proof isn't being verified in the smart contract. Need to use `verifyData()` to do this and also include the company signing `proof` from their DID document.

This is now documented and the existing VDAVerificationContract shows both of these steps working end to end.

Please review and let me know if you have any questions.

ITStar10 committed:
V 0.0.1 : Updated verification process

Updated
	Parameters of claimSBT()
	Verification process in claimSBT() using by new functions in VDA-Verification-Base
	Test code for claimSBT()

Added
	test/utils.ts : generate context for sign & request

tahpot commented on VDA-SBT/contracts/SoulboundNFT.sol:
```diff
@@ -190,57 +164,58 @@ contract SoulboundNFT is VDAVerificationContract,
+                // "-",
+                sbtInfo.recipient);
+
             params = abi.encodePacked(
```
Wrapping the `params` up twice is very confusing and seems unnecessary. I found this very difficult to follow.

ITStar10 replied:
This is because, "stack too deep" issue in the solidity.
Can't encode all the parameters with one `encodePacked()` function.

ITStar10 replied:
That is the limit of Solidity.

tahpot replied:
Oh :(

Where can I read more about that limit?

tahpot commented on VDA-SBT/contracts/SoulboundNFT.sol:
```diff
@@ -190,57 +164,58 @@ contract SoulboundNFT is VDAVerificationContract,
+            
+            verifyRequest(did, params, requestSignature, requestProof);
+
+            verifyData(bytes(sbtInfo.uniqueId), sbtInfo.signedData, sbtInfo.signedProof);
```
See the original requirements here: https://github.com/verida/blockchain-contracts/issues/57#issuecomment-1298148464

The `proofString` (which you have called `uniqueId`) is in this format:

`twitter-${twitterId}-${did}`

As such, you need to build that proof string in the smart contract:

const proofString = `${sbtType}-${uniqueId}-${did}` and call `verifyData()` on that:

verifyData(proofString, sbtInfo.signedData, sbtInfo.signedProof);

You'll need to update L186 in the test to generate the correct `signedData` on the client side.

ITStar10 replied:
Let's remove "-" symbols in the proof too as we removed them in the request params.

tahpot replied:
Ok

ITStar10 committed:
Minor updates for signer message

tahpot reviewed (changes requested):
@ITStar10 This is looking better.

After speaking to the team, we need to modify the restriction around minting. We currently restrict to only mint once per sbtType, but this needs to change to be once per uniqueId.

Currently we have:

```
require(_userInfo[sbtInfo.recipient][sbtInfo.sbtType] == 0, "Already claimed type");
```

This needs to change to be:

```
require(_userInfo[sbtInfo.recipient][sbtInfo.uniqueId] == 0, "Already claimed type");
```

ITStar10 committed:
Updated on Review

Commit:
Add missing verida dependencies

Commit:
Update endpoints to use devnet

ITStar10 committed:
Enhancement

- Updated
	getClaimedSBTList() returns tokenID list
	tokenInfo() returns SBTType & UniqueID

- Added
	burnSBT() implemented

tahpot commented:
Thanks @ITStar10, this looks good to merge.

Can you also deploy to Mumbai and add to meta transaction server?

ITStar10 committed:
Deploy script updated

Updated
	scripts/deploy.ts for deploy

Added
	yarn scripts to the package.json file
